libcloud/storage/drivers/azure_blobs.py

```diff
@@ -825,7 +826,12 @@ def upload_object_via_stream(self, iterator, container, object_name,
         """
         self._check_values(ex_blob_type, ex_page_blob_size)
         if ex_blob_type == "BlockBlob":
-            iterator.seek(0, os.SEEK_END)
+            try:
+                iterator.seek(0, os.SEEK_END)
+            except AttributeError:
+                buffer = BytesIO()
+                buffer.writelines(iterator)
```

Comment:
How does this work - does it need to exhaust and buffer the whole iterator in memory (inside buffer object) or does it read the iterator as it iterates over it and writes it as part of a request?

Reply:
Yes, this does buffer the whole iterator in memory unfortunately.

Given that [the content-size header is required](https://docs.microsoft.com/en-us/rest/api/storageservices/put-blob#request-headers-all-blob-types) we have to find the size of the iterator before making the request so I don't see a way to avoid this for the general case. E.g. using [tee](https://docs.python.org/3/library/itertools.html#itertools.tee) followed by something like [ilen](https://more-itertools.readthedocs.io/en/stable/_modules/more_itertools/more.html#ilen) will still copy the iterator in memory.

One potential work-around could be to upload chunks from the iterator via individual [put block](https://docs.microsoft.com/en-us/rest/api/storageservices/put-block) requests followed with a [put block list](https://docs.microsoft.com/en-us/rest/api/storageservices/put-block-list) request, but that would be a much more invasive change to the codebase. (Note that I also haven't proved out this approach in code yet so for now it's just a hypothesis from reading the docs.)

Do you have any suggestions to improve this?

Reply:
I'm fine with this for now, but at the very least, we should add a note to the documentation (and docstring) which points out that using that method will buffer the whole object in memory.

I do think that using put block would be a good future improvement (especially for very large objects). IIRC, one of the existing storage drivers already utilizes tsimilar approach, but I can't recall which one (it does require a bit more work though to track all the chunks, etc.).

Reply:
Added a note to the docstring for now.